### PR TITLE
VLLM backend support

### PR DESCRIPTION
**Pull request by NeonBohdan**
# Description
Added support for
- `ask` using beam search
- `vote` using perplexity score
- vllm openai compatible backend
- reproducability with hf demo checked 

# Other Notes
`Chat completions` don't support prompt logprobs
Using `complitions` for now
But once supported better migrate to `chat completions`

### CODE REVIEW

File: neon_llm_vllm/vllm.py [hunk 1]
```diff
@@ -134,24 +175,78 @@ def _score(self, prompt: str, targets: List[str], persona: dict) -> List[float]:
...
+            if start_index is None and start_pos < next_pos:
+                start_index = i
+
+            if end_index is None and end_pos <= next_pos:
+                end_index = i
+                break
+            
+            current_pos = next_pos
+
+        return start_index, end_index
```
Comment (NeonKirill) [inline neon_llm_vllm/vllm.py]: add trailing space

File: neon_llm_vllm/vllm.py [hunk 2]
```diff
@@ -134,24 +175,78 @@ def _score(self, prompt: str, targets: List[str], persona: dict) -> List[float]:
...
+            max_tokens=1,
+            temperature=0,
+            extra_body={
+                "skip_special_tokens": False
+            }
+        )
+        logprobs_list = [choice.logprobs for choice in completion.choices]
+        return logprobs_list
+
+    def _ppl(self, answer: str, logprobs: openai.types.completion_choice.Logprobs) -> float:
```
Comment (NeonKirill) [inline neon_llm_vllm/vllm.py]: rename to something like `_compute_ppl_score`

File: neon_llm_vllm/vllm.py [hunk 3]
```diff
@@ -89,24 +121,29 @@ def get_sorted_answer_indexes(self, question: str, answers: List[str], persona:
...
         """
 
-        response = openai.ChatCompletion.create(
+        response = self.model.completions.create(
             model=self.llm_model_name,
-            messages=prompt,
+            prompt=prompt,
             temperature=0,
             max_tokens=self.max_tokens,
+            extra_body={
```
Comment (NeonKirill) [inline neon_llm_vllm/vllm.py]: maybe set these properties in `__init__` with ability to override them from configs

File: neon_llm_vllm/vllm.py [hunk 4]
```diff
@@ -41,38 +42,69 @@ class ChatGPT(NeonLLM):
...
         return self._tokenizer
 
     @property
     def tokenizer_model_name(self) -> str:
-        return ""
+        return self.model_name
 
     @property
-    def model(self) -> openai:
+    def model(self) -> openai.OpenAI:
```
Comment (NeonKirill) [inline neon_llm_vllm/vllm.py:92]: please use `@cached_property` instead

File: neon_llm_vllm/vllm.py [hunk 5]
```diff
@@ -41,38 +42,69 @@ class ChatGPT(NeonLLM):
...
+    def context_depth(self):
+        return self._context_depth
+
+    @context_depth.setter
+    def context_depth(self, value):
+        self._context_depth = value + value % 2
+
     @property
     def tokenizer(self) -> None:
+        if self._tokenizer is None:
```
Comment (NeonKirill) [inline neon_llm_vllm/vllm.py]: please use @cached_property instead

File: neon_llm_vllm/rmq.py [hunk 6]
```diff
@@ -39,12 +39,12 @@ def __init__(self):
 
     @property
     def name(self):
-        return "chat_gpt"
+        return "vllm"
 
     @property
     def model(self):
```
Comment (NeonKirill) [inline neon_llm_vllm/rmq.py]: please use @cached_property instead

File: neon_llm_vllm/vllm.py [hunk 7]
```diff
@@ -41,38 +42,69 @@ class ChatGPT(NeonLLM):
...
+        _ = self.tokenizer
+
+    def get_metadata(self):
+        models = self.model.models.list()
+        vllm_model_name = models.data[0].id
+
+        model_name, *suffix = vllm_model_name.split("@")
+        revision = dict(enumerate(suffix)).get(0, None)
+
+        self.vllm_model_name = vllm_model_name
```
Comment (NeonKirill) [inline neon_llm_vllm/vllm.py]: please initialize attributes either in `__init__` or as `@property`

File: requirements/requirements.txt [hunk 8]
```diff
@@ -1,5 +1,7 @@
 # model
-openai[embeddings]~=0.27
+openai~=1.37
+transformers[sentencepiece]~=4.42
+jinja2>=3.1.0
```
Comment (NeonKirill) [inline requirements/requirements.txt]: please freeze particular version so future releases of Jinja won't break it

**Review by NeonKirill — CHANGES_REQUESTED** [review]

File: requirements/requirements.txt (hunk 8 above)
Comment (NeonBohdan) [reply requirements/requirements.txt]: Can be changed to `jinja2~=3.1.0`

**Pull request merged**